paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonSparkWriter.scala

```diff
@@ -229,10 +255,27 @@ case class PaimonSparkWriter(table: FileStoreTable) {
         writeWithoutBucket(data)
 
       case HASH_FIXED =>
-        // Topology: input -> bucket-assigner -> shuffle by partition & bucket
-        writeWithBucketProcessor(
-          withInitBucketCol,
-          CommonBucketProcessor(table, bucketColIdx, encoderGroupWithBucketCol))
+        if (!paimonExtensionEnabled) {
+          // Topology: input -> bucket-assigner -> shuffle by partition & bucket
+          writeWithBucketProcessor(
+            withInitBucketCol,
+            CommonBucketProcessor(table, bucketColIdx, encoderGroupWithBucketCol))
+        } else {
+          // Topology: input -> shuffle by partition & bucket
+          val bucketNumber = table.coreOptions().bucket()
+          val bucketKeyCol = tableSchema
+            .bucketKeys()
+            .asScala
+            .map(tableSchema.fieldNames().indexOf(_))
+            .map(x => col(data.schema.fieldNames(x)))
+            .toSeq
+          val args = Seq(lit(bucketNumber)) ++ bucketKeyCol
+          val repartitioned =
```

Comment:
what about  `data.withColumn(BUCKET_COL, call_udf(BucketExpression.FIXED_BUCKET, args: _*))`, so that we can use `iter.foreach(row => write.write(row, row.getInt(bucketColIdx)))` to avoid computing bucket id twice